File: testers/oracle/java/src/java/oracledb/OracleTestHelper.java
package oracledb;

import dev.typr.foundations.Connection;
import dev.typr.foundations.ConnectionRead;
import dev.typr.foundations.SqlConsumer;
import dev.typr.foundations.SqlFunction;
import dev.typr.foundations.Transactor;
import dev.typr.foundations.connect.OracleConfig;

public class OracleTestHelper {
  private static final OracleConfig CONFIG =
      OracleConfig.builder("localhost", 1521, "FREEPDB1", "typr", "typr_password")
          .serviceName("FREEPDB1")
          .build();

  private static final Transactor TRANSACTOR = Transactor.create(CONFIG).rollbackOnly();

  public static <T> T apply(SqlFunction<Connection, T> f) {
    return TRANSACTOR.transact(f);
  }

  public static void run(SqlConsumer<Connection> f) {
    TRANSACTOR.transactVoid(f);
  }

  public static <T> T applyRead(SqlFunction<ConnectionRead, T> f) {
    return TRANSACTOR.transactRead(f);
  }
}

File: testers/oracle/kotlin/src/kotlin/oracledb/OracleTestHelper.kt
package oracledb

import dev.typr.foundationskt.Connection
import dev.typr.foundationskt.ConnectionRead
import dev.typr.foundationskt.Transactor
import dev.typr.foundationskt.connect.OracleConfig

object OracleTestHelper {
    private val CONFIG: OracleConfig =
        OracleConfig.builder("localhost", 1521, "FREEPDB1", "typr", "typr_password")
            .serviceName("FREEPDB1")
            .build()

    private val TRANSACTOR: Transactor = Transactor.create(CONFIG).rollbackOnly()

    fun <T> apply(f: (Connection) -> T): T = TRANSACTOR.transact(f)

    fun run(f: (Connection) -> Unit) = TRANSACTOR.transact { conn -> f(conn) }

    fun <T> applyRead(f: (ConnectionRead) -> T): T = TRANSACTOR.transactRead(f)
}
